### PR TITLE
Scope initial swift

### PR DESCRIPTION
**Pull request by jayahariv**

### CODE REVIEW

File: Objective-C/CBLCollection.h [hunk 1]
```diff
@@ -19,21 +19,59 @@
...
+ 
+ When creating a new collection, the collection name, and the scope name are required.
+ The naming rules of the collections and scopes are as follows:
+ - Must be between 1 and 251 characters in length.
+ - Can only contain the characters A-Z, a-z, 0-9, and the symbols _, -, and %.
+ - Cannot start with _ or %.
+ - Both scope and collection names are case sensitive.
+ 
+ ##  `CBLCollection` Lifespan
+ `CBLCollection` is ref-counted and is owned by the database object that creates it. Hence,
```
Comment (pasin) [inline Objective-C/CBLCollection.h:55]: This is not related to Objective-C as user will not do call retain() or release(). Can you revise this?

File: Objective-C/CBLDatabase.h [hunk 2]
```diff
@@ -493,25 +494,14 @@ __deprecated_msg("Use [[database defaultCollection] setDocumentExpirationWithID:
...
-
 #pragma mark -- Scopes
 
 /**
  Get scope names that have at least one collection.
  
  @note: The default scope is exceptional as it will always be listed even though there are
  no collections under it. */
-- (NSArray*) scopes;
+- (NSArray<CBLScope*>*) scopes;
```
Comment (pasin) [inline Objective-C/CBLDatabase.h:504]: nullable and could return error now.

File: Objective-C/CBLDatabase.h [hunk 3]
```diff
@@ -530,7 +520,7 @@ __deprecated_msg("Use [[database defaultCollection] getDocumentExpirationWithID:
  @param scope Scope name
  @return list of collections in the scope.
  */
-- (NSArray*) collections: (nullable NSString*)scope;
+- (NSArray<CBLCollection*>*) collections: (nullable NSString*)scope;
```
Comment (pasin) [inline Objective-C/CBLDatabase.h:523]: nullable and could return error now.

File: Objective-C/CBLDatabase.h (hunk 2 above)
Comment (pasin) [reply Objective-C/CBLDatabase.h:504]: Please check the latest email about API change.

Here at the list of API that could return an error.
```
**Collection (Section 5.2, 5.5)**
Indexes() -> [String]
getDocument(id: String) -> Document?

**Scope (Section 5.4)**
getCollections() -> [Collection]
getCollection(name: String) -> Collection?
 
**Database (Section 5.6.2 , 5.6.3)**
getDefaultCollection() -> Collection?
getDefaultScope() -> Scope
getScopes() -> [Scope]
getScope(name: String) -> Scope?
getCollections(scope: String? = kDefaultScopeName) -> [Collection]
getCollection(name: String, scope: String? = kDefaultScopeName) -> Collection?
```

File: Objective-C/CBLDatabase.mm [hunk 4]
```diff
@@ -711,7 +717,7 @@ - (NSArray*) scopes {
     return [NSArray array];
 }
 
-- (nullable CBLScope*) scopeWithName: (nullable NSString*)name {
+- (nullable CBLScope*) scopeWithName: (NSString*)name {
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm:720]: Could return an error now.

File: Objective-C/CBLIndexable.h [hunk 5]
```diff
@@ -24,7 +24,7 @@ NS_ASSUME_NONNULL_BEGIN
 @protocol CBLIndexable <NSObject>
 
 /** Return all index names. */
-- (NSArray*) indexes;
+- (NSArray<NSString*>*) indexes;
```
Comment (pasin) [inline Objective-C/CBLIndexable.h:27]: Could return an error.

File: Objective-C/CBLReplicatorConfiguration.h [hunk 6]
```diff
@@ -239,6 +241,20 @@ __deprecated_msg("Use [... initWithTarget:] instead.");
 - (void) addCollections: (NSArray*)collections
                  config: (nullable CBLCollectionConfiguration*)config;
 
+/**
```
Comment (pasin) [inline Objective-C/CBLReplicatorConfiguration.h:244]: There is also a `collections` property.

File: Objective-C/CBLScope.h [hunk 7]
```diff
@@ -42,7 +42,7 @@ extern NSString* const kCBLDefaultScopeName;
  and the reference will remain valid until it's released. Most operations on the invalid
  CBLScope object will fail with null or empty result.
  */
-@interface CBLScope : NSObject
+@interface CBLScope : NSObject <CBLQueryFactory>
```
Comment (pasin) [inline Objective-C/CBLScope.h:45]: Scope doesn't implement CBLQueryFactory now. We discard this feature in helium.

File: Objective-C/CBLScope.h [hunk 8]
```diff
@@ -58,7 +58,7 @@ extern NSString* const kCBLDefaultScopeName;
 /**
  Get a collection in the scope by name.
  If the collection doesn't exist, a nil value will be returned. */
-- (CBLCollection*) collectionWithName: (NSString*)name;
+- (nullable CBLCollection*) collectionWithName: (NSString*)name;
```
Comment (pasin) [inline Objective-C/CBLScope.h:61]: Could return an error.

File: Objective-C/CBLScope.m [hunk 9]
```diff
@@ -45,4 +45,11 @@ - (CBLCollection *) collectionWithName: (NSString *)collectionName {
     return [NSArray array];
 }
 
+- (CBLQuery*) createQuery: (NSString*)query error: (NSError**)error {
```
Comment (pasin) [inline Objective-C/CBLScope.m:48]: Remove

File: Swift/Collection.swift [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+/// A `CBLCollection` represent a collection which is a container for documents.
```
Comment (pasin) [inline Swift/Collection.swift:22]: No CBL

File: Swift/Collection.swift [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+/// `kCBLDefaultCollectionName` and `kCBLDefaultScopeName` constant.
+///
+/// When creating a new collection, the collection name, and the scope name are required.
+/// The naming rules of the collections and scopes are as follows:
+/// - Must be between 1 and 251 characters in length.
+/// - Can only contain the characters A-Z, a-z, 0-9, and the symbols _, -, and %.
+/// - Cannot start with _ or %.
+/// - Both scope and collection names are case sensitive.
+///
+/// ## CBLCollection Lifespan
```
Comment (pasin) [inline Swift/Collection.swift:40]: This section is not valid for Swift. Please review this section.

File: Swift/Collection.swift [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+    /// The scope of the collection.
+    var scope: Scope { Scope(_impl.scope) }
+    
+    // MARK: Document Management
+    
+    /// Total number of documents in the collection.
+    var count: UInt64 { _impl.count }
+    
+    /// Get an existing document by id.
+    func document(id: String) -> Document? {
```
Comment (pasin) [inline Swift/Collection.swift:77]: Could throw

File: Swift/Collection.swift [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+            listener(CollectionChange(collection: self, documentIDs: change.documentIDs))
+        }
+        return ListenerToken(token)
+    }
+    
+    
+    // MARK: Indexable
+    
+    /// Return all index names
+    public func indexes() -> [String] {
```
Comment (pasin) [inline Swift/Collection.swift:232]: Could throw

File: Swift/Indexable.swift [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+
+public protocol Indexable {
+    // TODO: Remove this comment after code review!
+    // Since the protocol signature difference with
+    // 1. previous(var indexes) & current(func indexes() -> [String])
+    // 2. previous(createIndex(_ config: name: )) & current( createIndex(name: config:) ) arg reverse
+    // Database class will not implementthis protocol
+    
+    /// Return all index names
+    func indexes() -> [String]
```
Comment (pasin) [inline Swift/Indexable.swift:30]: Could throws

File: Swift/Indexable.swift [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+//
+
+import Foundation
+
+public protocol Indexable {
+    // TODO: Remove this comment after code review!
+    // Since the protocol signature difference with
+    // 1. previous(var indexes) & current(func indexes() -> [String])
+    // 2. previous(createIndex(_ config: name: )) & current( createIndex(name: config:) ) arg reverse
+    // Database class will not implementthis protocol
```
Comment (pasin) [inline Swift/Indexable.swift:27]: Agree

File: Swift/ReplicatorConfiguration.swift [hunk 16]
```diff
@@ -190,17 +200,71 @@ public struct ReplicatorConfiguration {
...
+    public func removeCollection(_ collection: Collection) -> ReplicatorConfiguration {
+        
+        // TODO: Add implementation
+        
+        return self
+    }
+    
+    /// Get a copy of the collection’s config. If the config needs to be changed for the collection, the
+    /// collection will need to be re-added with the updated config.
+    public func getCollectionConfig(_ collection: Collection) -> CollectionConfiguration {
```
Comment (pasin) [inline Swift/ReplicatorConfiguration.swift:261]: CollectionConfiguration is nullable.

File: Swift/Scope.swift [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+    
+    /// Scope name
+    public var name: String {
+        return _impl.name
+    }
+    
+    // MARK: Collections
+    
+    /// Get all collections in the scope.
+    public func collections() -> [Collection] {
```
Comment (pasin) [inline Swift/Scope.swift:35]: Could throws

File: Swift/Scope.swift [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+        var collections = [Collection]()
+        for c in _impl.collections() {
+            collections.append(Collection(impl: c))
+        }
+        
+        return collections
+    }
+    
+    /// Get a collection in the scope by name. If the collection doesn't exist, a nil value will be returned.
+    public func collection(name: String) -> Collection? {
```
Comment (pasin) [inline Swift/Scope.swift:45]: Could throws

File: Swift/Database.swift [hunk 19]
```diff
@@ -414,6 +430,87 @@ public final class Database {
...
+    public func defaultCollection() -> Collection? {
+        guard let c = _impl.defaultCollection() else {
+            return nil
+        }
+        
+        return Collection(impl: c)
+    }
+    
+    /// Get all collections in the specified scope.
+    public func collections(scope: String? = defaultScopeName) -> [Collection] {
```
Comment (pasin) [inline Swift/Database.swift:477]: Could throws

File: Swift/Database.swift [hunk 20]
```diff
@@ -414,6 +430,87 @@ public final class Database {
...
+            return nil
+        }
+        
+        return Scope(s)
+    }
+    
+    // MARK: Collections
+    
+    /// Get the default collection. If the default collection is deleted, null will be returned.
+    public func defaultCollection() -> Collection? {
```
Comment (pasin) [inline Swift/Database.swift:468]: Could throws

File: Swift/Database.swift [hunk 21]
```diff
@@ -414,6 +430,87 @@ public final class Database {
...
+            scopes.append(Scope(s))
+        }
+        
+        return scopes
+    }
+    
+    /// Get a scope object by name. As the scope cannot exist by itself without having a collection, the nil
+    /// value will be returned if there are no collections under the given scope’s name.
+    /// Note: The default scope is exceptional, and it will always be returned.
+    public func scope(name: String) -> Scope? {
```
Comment (pasin) [inline Swift/Database.swift:457]: Could throws

File: Swift/Database.swift [hunk 22]
```diff
@@ -414,6 +430,87 @@ public final class Database {
...
+    
+    /// Get the default scope.
+    public func defaultScope() -> Scope {
+        return Scope(_impl.defaultScope())
+    }
+    
+    /// Get scope names that have at least one collection.
+    /// Note: the default scope is exceptional as it will always be listed even though there are no collections
+    /// under it.
+    public func scopes() -> [Scope] {
```
Comment (pasin) [inline Swift/Database.swift:445]: Could throws

File: Swift/Database.swift [hunk 23]
```diff
@@ -414,6 +430,87 @@ public final class Database {
     /// Log object used for configuring console, file, and custom logger.
     public static let log = Log()
     
+    
+    // MARK: Scopes
+    
+    
+    /// Get the default scope.
+    public func defaultScope() -> Scope {
```
Comment (pasin) [inline Swift/Database.swift:438]: Could throws

File: Swift/Database.swift [hunk 24]
```diff
@@ -414,6 +430,87 @@ public final class Database {
...
+        if let err = error {
+            throw err
+        }
+        
+        return Collection(impl: result)
+    }
+
+    /// Get a collection in the specified scope by name.
+    /// If the collection doesn't exist, a nil value will be returned.
+    public func collection(name: String, scope: String? = defaultScopeName) -> Collection? {
```
Comment (pasin) [inline Swift/Database.swift:499]: Could throws

**Review by pasin — CHANGES_REQUESTED** [review]
The PR is big so I might not catch everything.

**Comment by pasin** [comment]
I just noticed that the PR was marked as draft, but please look at my review comment.

**Comment by jayahariv** [comment]
I waited for the checks to pass to make it ready, but forgot. Sure, it was ready, when I assigned.

File: Swift/Collection.swift [hunk 25]
```diff
@@ -0,0 +1,292 @@
...
+/// The naming rules of the collections and scopes are as follows:
+/// - Must be between 1 and 251 characters in length.
+/// - Can only contain the characters A-Z, a-z, 0-9, and the symbols _, -, and %.
+/// - Cannot start with _ or %.
+/// - Both scope and collection names are case sensitive.
+///
+/// ## CBLCollection Lifespan
+/// A `Collection` object and its reference remain valid until either
+/// the database is closed or the collection itself is deleted, in that case it will
+/// return CBLErrorNotOpen while accessing the collection APIs.
```
Comment (pasin) [inline Swift/Collection.swift:43]: `throw an NSError with the CBLError.notOpen code while ...`

File: Swift/Collection.swift [hunk 26]
```diff
@@ -0,0 +1,292 @@
...
+    var scope: Scope { Scope(_impl.scope) }
+    
+    // MARK: Document Management
+    
+    /// Total number of documents in the collection.
+    var count: UInt64 { _impl.count }
+    
+    /// Get an existing document by id.
+    ///
+    /// CBLErrorNotOpen code will be thrown if the collection is deleted
```
Comment (pasin) [inline Swift/Collection.swift:73]: CBLErrorNotOpen > An error with CBLError.notOpen code or `CBLError.notOpen code`.

Similarly for the other methods as well.

File: Swift/Collection.swift [hunk 27]
```diff
@@ -0,0 +1,292 @@
...
+    /// or the database is closed.
+    func setDocumentExpiration(id: String, expiration: Date?) throws {
+        try _impl.setDocumentExpirationWithID(id, expiration: expiration)
+    }
+    
+    /// Get the expiration date set to the document of the given id.
+    ///
+    /// CBLErrorNotOpen code will be thrown if the collection is deleted
+    /// or the database is closed.
+    func getDocumentExpiration(id: String) -> Date? {
```
Comment (pasin) [inline Swift/Collection.swift:214]: throws

File: Swift/DataSource.swift [hunk 28]
```diff
@@ -44,9 +44,15 @@ public final class DataSource {
     ///
     /// - Parameter database: The database object.
     /// - Returns: The database data source.
+    @available(*, deprecated, message: "Use database.defaultCollection().saveDocument(:) instead.")
```
Comment (pasin) [inline Swift/DataSource.swift:47]: Wrong message.

File: Swift/Database+Query.swift [hunk 29]
```diff
@@ -32,6 +32,7 @@ extension Database {
     }
     
     /// All index names.
+    @available(*, deprecated, message: "Use database.defaultCollection().indexes instead.")
```
Comment (pasin) [inline Swift/Database+Query.swift:35]: indexes()

File: Swift/Scope.swift [hunk 30]
```diff
@@ -0,0 +1,55 @@
...
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+public final class Scope {
```
Comment (pasin) [inline Swift/Scope.swift:22]: Add a description.

File: Objective-C/CBLScope.h [hunk 31]
```diff
@@ -18,7 +18,7 @@
 //
 
 #import <Foundation/Foundation.h>
-
+#import "CBLQueryFactory.h"
```
Comment (pasin) [inline Objective-C/CBLScope.h:21]: Remove

File: Objective-C/CBLScope.h [hunk 32]
```diff
@@ -52,13 +52,22 @@ extern NSString* const kCBLDefaultScopeName;
 
 #pragma mark Collections
 
-/** Get all collections in the scope. */
-- (NSArray<CBLCollection*>*) collections;
+/**
```
Comment (pasin) [inline Objective-C/CBLScope.h:55]: Please revise CBLScope Lifespan as it's not valid.

File: Swift/CollectionConfiguration.swift [hunk 33]
```diff
@@ -0,0 +1,43 @@
...
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+public struct CollectionConfiguration {
```
Comment (pasin) [inline Swift/CollectionConfiguration.swift:22]: Add a short description.

File: Swift/CollectionChangeObservable.swift [hunk 34]
```diff
@@ -0,0 +1,34 @@
...
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+public protocol CollectionChangeObservable {
```
Comment (pasin) [inline Swift/CollectionChangeObservable.swift:22]: Add a short description.

File: Swift/Indexable.swift [hunk 35]
```diff
@@ -0,0 +1,32 @@
...
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+public protocol Indexable {
```
Comment (pasin) [inline Swift/Indexable.swift:22]: Add a short description.

File: Swift/QueryFactory.swift [hunk 36]
```diff
@@ -0,0 +1,30 @@
...
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+public protocol QueryFactory {
```
Comment (pasin) [inline Swift/QueryFactory.swift:22]: Add a short description.

File: Objective-C/CBLCollectionChangeObservable.h [hunk 37]
```diff
@@ -17,12 +17,12 @@
...
 //
 
-#import "CBLDatabaseChange.h"
+#import "CBLCollectionChange.h"
 #import "CBLListenerToken.h"
 
 NS_ASSUME_NONNULL_BEGIN
 
-@protocol CBLDatabaseChangeObservable <NSObject>
+@protocol CBLCollectionChangeObservable <NSObject>
```
Comment (pasin) [inline Objective-C/CBLCollectionChangeObservable.h:25]: Add a short description.

File: Objective-C/CBLQueryFactory.h [hunk 38]
```diff
@@ -0,0 +1,39 @@
...
+//  limitations under the License.
+//
+
+#import <Foundation/Foundation.h>
+
+@class CBLQuery;
+
+NS_ASSUME_NONNULL_BEGIN
+
+@protocol CBLQueryFactory <NSObject>
```
Comment (pasin) [inline Objective-C/CBLQueryFactory.h:26]: Add a short description.

**Review by pasin — CHANGES_REQUESTED** [review]

**Review by pasin — APPROVED** [review]

**Pull request merged**